ws2812b/src/ws2812b_lld.c

```diff
@@ -1,363 +1,394 @@
-        "nop                                 \n"// H t11
-        "nop                                 \n"// H t12
-        RGBLedPin_ReSet                         // L t1
-        "DECW    Y                           \n"//2
-        "JREQ    L:lb_exit                   \n"//1 on loop, 2 jmp to exit
-        "JP      L:lb_begin_loop             \n"//5
-                                                // 
-        "lb_bit7_Send_0:                     \n"//------Send 0 : 400ns High, 850ns Low (6,14) 375ns,875ns
-        RGBLedPin_Set                           // H t1
-        "nop                                 \n"// H t2
-        "nop                                 \n"// H t3
-        "nop                                 \n"// H t4
-        "nop                                 \n"// H t5
-        "nop                                 \n"// H t6
-        RGBLedPin_ReSet                         // L t1
-        "nop                                 \n"// L t1
-        "nop                                 \n"// L t2
-        "nop                                 \n"// L t3
-        "nop                                 \n"// L t4
-        "nop                                 \n"// L t5
-        "nop                                 \n"// L t6
-                                                // L DECW 2, JREQ 1, 2 = 5 =>   t14
-//--------------------------------------------------------
-//--------------------------------------------------------
-        "DECW    Y                           \n"//2
-        "JREQ    L:lb_exit                   \n"//1 on loop, 2 jmp to exit
-        "JP      L:lb_begin_loop             \n"//5
-        "lb_exit:nop");
-  
-    __set_interrupt_state(int_state);
-}
\ No newline at end of file
+/** 
+  ******************************************************************************
+  * @file    ws2812b_lld.c
+  * @brief   Low Level Driver for WS2812B Type RGB/RGBW LED strips
+  * @author  https://github.com/j3qq4hch
+  * @author  https://github.com/mnemocron
+  * @date    2020-10-27
+  * 
+  * @note    IAR Embedded Workbench 3.11.1
+  *          shared components 8.3.2.5988
+  * @note    tested with RGBW LED type: IN-PI55QATPRPGPBPW-XX
+  * 
+  ******************************************************************************
+  */
+
+/**@todo is this include file necessary? it compiles without it */
+// #include <iostm8s103f3.h>
+#include <intrinsics.h>
+#include "ws2812b_lld.h"
+
+// depending on RGBW or RGB type it is 3 or 4 bytes to manage
+#ifdef WS2812B_MODE_RGBW
+  #define COL_PER_LED (4)
+
+  const RGBColor_t RED            = {255,   0,   0,   0};
+  const RGBColor_t GREEN          = {  0, 255,   0,   0};
+  const RGBColor_t BLUE           = {  0,   0, 255,   0};
+  const RGBColor_t YELLOW         = {255, 255,   0,   0};
+  const RGBColor_t MAGENTA        = {255,   0, 255,   0};
+  const RGBColor_t BLACK          = {  0,   0,   0,   0};
+  const RGBColor_t WHITE          = {255, 255, 255,   0};

```

Comment:
LLD file is about dealing with WS2812 protocol and hardware. Colors definitions should be placed to separate file then

Reply:
So now you are reviewing the original source code, not the changes in the my PR.
Thanks a lot, I am learning a lot today :)
While we are at it, can you recommend any literature on "good practices in embedded software design"?
Where you learn e.g.
- which code to separate into files / libraries
- generally how to build robust `ifdef` structures
- when to use/avoid certain data types

Reply:
Yeah, I write comments because I spent quite a bit on this project, and I'm glad to share knowledge. I think you don't need to correct your PR, I'm going to make the changes myself. About literature - I'm an old guy, so books I read decades ago, may be not good for you )
These matters are somewhat advanced and you get it with experience. Reading others code, etc. Sorry I don't know any modern C/C++ book for embedded. I often work with Linux source code which is almost all plain C and some C++. But Linux is too complex for you (and me)), but keep your eye on RTOS (FreeRTOS, works for example on STM32, and STM supports it officially. It is part of Expressiv SDK ESP32 BTW).
You're probably working with baremetal code, next step is RTOS when you'll get out of sandbox.

Reply:
I appreciate the feedback! We have briefly touched Linux and RTOS in school where the focus was on the advantages and the new special mechanisms (threads, semaphores, mutexes etc.) but never about how to organize code in a project. So I have my own style in STM32 projects but never quite know if it is good or bad practice ¯\_(ツ)_/¯

Reply:
Actually my own code style is far from ideal too. Ability for learning is the greatest skill I know)

Reply:
I've found this STM8 port of FastLED library: https://github.com/thielj/FastLED-STM8, I'm going to try it. Seems to be using IAR compiler